File: android/src/main/java/com/margelo/nitro/nitroimagepipeline/transform/FitGeometry.kt
package com.margelo.nitro.nitroimagepipeline.transform

import com.margelo.nitro.nitroimagepipeline.ResizeFit
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * The resize geometry both platforms implement — a port of `FitGeometry` in
 * `ios/FitResizeProcessor.swift`; `example/__tests__/fit-geometry.ts` is the copy the harness
 * checks both against. With a `w` × `h` source and a `W` × `H` box:
 * ```
 * cover   : sx = sy = max(W/w, H/h)
 * contain : sx = sy = min(W/w, H/h)
 * stretch : sx = W/w ; sy = H/h
 * center  : sx = sy = 1
 * if !allowUpscale: sx = min(sx, 1) ; sy = min(sy, 1)
 * scaled  = (round(w·sx), round(h·sy))
 * output  = (min(W, scaled.w), min(H, scaled.h))
 * origin  = trunc((output − scaled) / 2)        — a centred, whole-pixel crop
 * ```
 *
 * The source is drawn scaled to `scaled` at `origin` on an `output`-sized canvas. Rounding is
 * half-away-from-zero (`roundToInt()` here, `.rounded()` in Swift — identical for the positive
 * values involved), and the divisions are done in the same order in `Double` on both sides so the
 * two agree to the pixel.
 */
internal data class FitGeometry(
    val sourceWidth: Int,
    val sourceHeight: Int,
    val scaledWidth: Int,
    val scaledHeight: Int,
    val outputWidth: Int,
    val outputHeight: Int,
    val originX: Int,
    val originY: Int,
) {
  /**
   * Whether the source is already the output — nothing to draw. Both sizes are compared with the
   * *source*: `scaled == output` alone also holds for every plain downscale (nothing is cropped),
   * which is not an identity.
   */
  val isIdentity: Boolean
    get() =
        scaledWidth == sourceWidth &&
            scaledHeight == sourceHeight &&
            outputWidth == sourceWidth &&
            outputHeight == sourceHeight

  companion object {
    fun of(
        sourceWidth: Int,
        sourceHeight: Int,
        boxWidth: Int,
        boxHeight: Int,
        fit: ResizeFit,
        allowUpscale: Boolean,
    ): FitGeometry {
      val source = sourceWidth.toDouble() to sourceHeight.toDouble()
      val box = boxWidth.toDouble() to boxHeight.toDouble()
      var scaleX: Double
      var scaleY: Double
      when (fit) {
        ResizeFit.COVER -> {
          val scale = max(box.first / source.first, box.second / source.second)
          scaleX = scale
          scaleY = scale
        }
        ResizeFit.CONTAIN -> {
          val scale = min(box.first / source.first, box.second / source.second)
          scaleX = scale
          scaleY = scale
        }
        ResizeFit.STRETCH -> {
          scaleX = box.first / source.first
          scaleY = box.second / source.second
        }
        ResizeFit.CENTER -> {
          scaleX = 1.0
          scaleY = 1.0
        }
      }
      if (!allowUpscale) {
        scaleX = min(scaleX, 1.0)
        scaleY = min(scaleY, 1.0)
      }
      val scaledWidth = (source.first * scaleX).roundToInt().coerceAtLeast(1)
      val scaledHeight = (source.second * scaleY).roundToInt().coerceAtLeast(1)
      val outputWidth = min(boxWidth, scaledWidth)
      val outputHeight = min(boxHeight, scaledHeight)
      return FitGeometry(
          sourceWidth = sourceWidth,
          sourceHeight = sourceHeight,
          scaledWidth = scaledWidth,
          scaledHeight = scaledHeight,
          outputWidth = outputWidth,
          outputHeight = outputHeight,
          // Integer division truncates toward zero, like `.rounded(.towardZero)` on iOS.
          originX = (outputWidth - scaledWidth) / 2,
          originY = (outputHeight - scaledHeight) / 2,
      )
    }
  }
}

File: example/__tests__/fit-geometry.ts
import type { ResizeFit } from 'react-native-nitro-image-pipeline';

/**
 * The resize geometry the native sides implement — `FitGeometry` in
 * `ios/FitResizeProcessor.swift` and `android/.../transform/FitGeometry.kt`.
 * The suites compute their expected sizes from this copy, so one table
 * checks both platforms against the same contract:
 * ```
 * cover   : sx = sy = max(W/w, H/h)
 * contain : sx = sy = min(W/w, H/h)
 * stretch : sx = W/w ; sy = H/h
 * center  : sx = sy = 1
 * if !allowUpscale: sx = min(sx, 1) ; sy = min(sy, 1)
 * scaled  = (round(w·sx), round(h·sy))
 * output  = (min(W, scaled.w), min(H, scaled.h))
 * ```
 * `Math.round` rounds halves up, like Kotlin's `roundToInt()` and Swift's
 * `.rounded()` do for the positive values involved.
 */
export function fitGeometry(
  sourceWidth: number,
  sourceHeight: number,
  boxWidth: number,
  boxHeight: number,
  fit: ResizeFit,
  allowUpscale = true,
): { width: number; height: number } {
  const W = Math.round(boxWidth);
  const H = Math.round(boxHeight);
  let sx: number;
  let sy: number;
  switch (fit) {
    case 'cover':
      sx = sy = Math.max(W / sourceWidth, H / sourceHeight);
      break;
    case 'contain':
      sx = sy = Math.min(W / sourceWidth, H / sourceHeight);
      break;
    case 'stretch':
      sx = W / sourceWidth;
      sy = H / sourceHeight;
      break;
    case 'center':
      sx = sy = 1;
      break;
  }
  if (!allowUpscale) {
    sx = Math.min(sx, 1);
    sy = Math.min(sy, 1);
  }
  const scaledWidth = Math.max(Math.round(sourceWidth * sx), 1);
  const scaledHeight = Math.max(Math.round(sourceHeight * sy), 1);
  return {
    width: Math.min(W, scaledWidth),
    height: Math.min(H, scaledHeight),
  };
}
